### DISCUSSION TITLE
Assignment - Design a parking lot

### DISCUSSION THREAD

**Discussion by kanmaytacker** [kanmaytacker/design-questions#7] [General]
What?

The objective is to come up with the design and implementation of a parking lot.
Follow the worksheet [here](https://github.com/kanmaytacker/design-questions/blob/master/notes/04-design-a-parking-lot-worksheet.md) to get started.

Why?
Popular LLD question.
Simple enough to implement end to end.

How?
1. Identify the classes to be needed. Add them [here](https://github.com/kanmaytacker/design-questions/blob/master/notes/04-design-a-parking-lot-worksheet.md#class-diagram)
2. Create a class diagram for the above classes.
3. Identify the APIs to be created from the use case diagram.

**Comment by kanmaytacker** [comment]
Submission by [Akash](https://github.com/akash-bhardwaj).

* [Class Diagram](https://github.com/kanmaytacker/design-questions/discussions/6)
* APIs
  * [Admin APIs](https://github.com/kanmaytacker/design-questions/discussions/4)
  * [Parking Attendant APIs ](https://github.com/kanmaytacker/design-questions/discussions/5)

**Reply by Faizahmadfaiz** [reply]
### Class Diagram
```mermaid
classDiagram

class ParkingSpotType {
   <<enumeration>>
   SMALL, MEDIUM, LARGE
}

class ParkingSpotStatus {
   <<enumeration>>
   VAILABLE, OCCUPIED, OUT_OF_SERVICE
}

class PaymentType {
   <<enumeration>>
   CASH, ONLINE, CARD
}

class VehicleType {
   <<enumeration>>
   TWO_WHEELER, THREE_WHEELER, FOUR_WHEELER, HEAVY_DUTY
}

class Gate {
   -ParkingAttendant attendant
   
   +pay(Ticket)
   +issueTicket(Vehicle)
}

Gate --o ParkingAttendant

class Vehicle {
   String vehicleNumber
   VehicleType vehicleType
}

class ParkingAttendant {
   -Ticket ticket
   -Payment payment
   -List<ParkingSpot> parkingSpots
   -SpotAllocationStrategy spotAllocationStrategy
   +receievePayment(Ticket)
   +issueTicket(Vehicle)
   +allocateSpot()
}
ParkingAttendant --o Ticket
ParkingAttendant --o Payment
ParkingAttendant --o ParkingSpot
ParkingAttendant --o SpotAllocationStrategy

class SpotAllocationStrategy {
   <<interface>>
   +allocateSpot(ParkingSpot[])
}

class RandomAllocationStrategy {
   +allocateSpot(ParkingSpot[])
}

class NearestToGateAllocationStrategy {
   +allocateSpot(ParkingSpot[])
}

SpotAllocationStrategy <-- RandomAllocationStrategy : implements
SpotAllocationStrategy <-- NearestToGateAllocationStrategy : implements
ParkingAttendant --o SpotAllocationStrategy

class Ticket {
   -ParkingSpot spot
   -String ticketId
   -Vehicle vehicle
   -DateTime entryTime
   -Double price
   
   +issue() void
}

Ticket --* Vehicle

class ParkingSpot {
   -int spotNumber
   -ParkingSpotType type
   -ParkingSpotStatus status
   -int floor
   
   +allot() void
   +deallot() void
}


class Payment {
   -PaymentType type
   
   +makePayment(Ticket)
}

Ticket --o ParkingSpot
```

### Api Endpoints
#### CRUD for Parking lot (Related to Admin)
1. Add Parking lot
   - ```/parkinglot```
   - POST
   - BODY: ```{name: string, city: string, country: string, zip: number, floors: number, ...other employee details}```
2. Update Parking lot 
   - ```/parkinglot/:parkinglotId```
   - PUT
   - BODY: same as above
3. Get all Parking lots
    - ```/parkinglot```
    - GET
4. Get a parking lot
    - ```/parkinglot/:parkinglotId```
    - GET
5. Delete a parking lot
    - ```/parkinglot/:parkinglotId```
    - DELETE

#### CRUD For Parking slots (ADMIN)
1. Add a parking spot
    - ``/parkinglot/:parkinglotId/parkingspot``
    - POST
    - BODY: ``type: PARKINGSPOTTYPE, status: AVAILABLE. floor: number``
2. See Details of all parking spots(Used for dashboard purpose, available for attendant and customers)
   - ``/parkinglot/:parkinglotId/parkingspot``
   - GET
3. Check status for a parking spot (For Parking attendant also)
   - ``/parkinglot/:parkinglotId/parkingspot/:parkingspotId``
4. Update a parking spot (for attendant also)
   - ``/parkinglot/:parkinglotId/parkingspot/:parkingspotId``
   - PUT
5. Delete a parking spot
    - ``/parkinglot/:parkinglotId/parkingspot/:parkingspotId``
    - DELETE
   
#### Related to Customers
1. Check parking spot's status (ABOVE)
2. Make Payment
    - ``parkinglot/:parkinglotId/:ticketId/pay``
    - POST/GET ?
   
#### Related to Parking Attendant
1. Create a ticket
    - ``parkinglot/:parkinglotId/ticket``
    - POST
    - BODY ``vehicleNumber: string, entryTime: string, vehicleType: VEHICLETYPE, paymentStatus: NOT PAID, parkingSpot: int``
2. Read a ticket
    - ``parkinglot/:parkinglotId/ticket``
    - GET

**Comment by kanmaytacker** [comment]
I have updated the answers sheet [here](https://github.com/kanmaytacker/design-questions/blob/master/notes/04-design-a-parking-lot-answers.md#class-diagram) with the class diagram.
Next steps:
* Identify the differences between the two designs.
* Identify problems and possible solutions in the above design
* List the APIs required for your parking lot system

**Reply by Sowndarya011** [reply]
![Parking Lot - class diagram](https://user-images.githubusercontent.com/115414728/198569974-dfb5755f-dc82-433f-a33c-c45e3c840db0.png)
Class diagram for Parking Lot Problem.

**Comment by kanmaytacker** [comment]
I have merged the API design on the [master branch ](https://github.com/kanmaytacker/design-questions/blob/master/notes/04-design-a-parking-lot-answers.md#api-design)as well.
